lib/babe/parachain_inherents.go

```diff
@@ -0,0 +1,240 @@
+
+// Index Returns VDT index
+func (im Implicit) Index() uint {
+	return 1
+}
+
+// Explicit is for explicit attestation.
+type Explicit ValidatorSignature
+
+// Index Returns VDT index
+func (ex Explicit) Index() uint {
+	return 2
+}
+
+// GetValidityAttestation returns a ValidityAttestation varying data type.
+func GetValidityAttestation() ValidityAttestation {
+	vdt, err := scale.NewVaryingDataType(Implicit{}, Explicit{})
+	if err != nil {
+		panic(err)
+	}
+
+	return ValidityAttestation(vdt)
+}
+
+// DisputeStatement is a statement about a candidate, to be used within the dispute
+// resolution process. Statements are either in favour of the candidate's validity
+// or against it.
+type DisputeStatement scale.VaryingDataType
+
+// Valid is for a valid statement
+type Valid ValidDisputeStatementKind
+
+// Index Returns VDT index
+func (v Valid) Index() uint {
+	return 0
+}
+
+// Invalid is for an invalid statement
+type Invalid InvalidDisputeStatementKind
+
+// Index Returns VDT index
+func (in Invalid) Index() uint {
+	return 1
+}
+
+// ValidDisputeStatementKind is a kind of statements of validity on a candidate.
+type ValidDisputeStatementKind scale.VaryingDataType
+
+// ExplicitValidDisputeStatementKind is an explicit statement issued as part of a dispute.
+type ExplicitValidDisputeStatementKind struct{}
+
+// Index Returns VDT index
+func (ex ExplicitValidDisputeStatementKind) Index() uint {
+	return 0
+}
+
+// BackingSeconded is a seconded statement on a candidate from the backing phase.
+type BackingSeconded common.Hash
+
+// Index Returns VDT index
+func (bs BackingSeconded) Index() uint {
+	return 1
+}
+
+// BackingValid is a valid statement on a candidate from the backing phase.
+type BackingValid common.Hash
+
+// Index Returns VDT index
+func (bv BackingValid) Index() uint {
+	return 2
+}
+
+// ApprovalChecking is an approval vote from the approval checking phase.
+type ApprovalChecking struct{}
+
+// Index Returns VDT index
+func (ac ApprovalChecking) Index() uint {
+	return 3
+}
+
+// InvalidDisputeStatementKind is a kind of statements of invalidity on a candidate.
+type InvalidDisputeStatementKind scale.VaryingDataType
+
+// ExplicitInvalidDisputeStatementKind is an explicit statement issued as part of a dispute.
+type ExplicitInvalidDisputeStatementKind struct{}
+
+// Index Returns VDT index
+func (ex ExplicitInvalidDisputeStatementKind) Index() uint {
+	return 0
+}
+
+// GetDisputeStatement returns a GetDisputeStatement varying data type.
+func GetDisputeStatement() DisputeStatement {
+	invalidDisputeStatementKind, err := scale.NewVaryingDataType(ExplicitInvalidDisputeStatementKind{})
+	if err != nil {
+		panic(err)
+	}
+
+	validDisputeStatementKind, err := scale.NewVaryingDataType(
+		ExplicitValidDisputeStatementKind{}, BackingSeconded{}, BackingValid{}, ApprovalChecking{})
+	if err != nil {
+		panic(err)
+	}
+
+	vdt, err := scale.NewVaryingDataType(validDisputeStatementKind.Value(), invalidDisputeStatementKind.Value())
```

Comment:
I don't think this is right.  I think you need to implement the `scale.VaryingDataTypeValue` interface  on the `ValidDisputeStatementKind` and `InvalidDisputeStatementKind` since this is a nested enum.

Reply:
`DisputeStatement` has two values `Valid` (0) and `Invalid` (1). So, I have structs Valid and Invalid which implement Index() method (thus scale.VaryingDataTypeValue interface). ValidDisputeStatementKind is type of Valid and InvalidDisputeStatementKind is type of Invalid. Isn't that appropriate?

Checkout 
https://github.com/ChainSafe/gossamer/pull/2566/files#diff-3abb401d0f0c9e3a6c9f813400f674af4b2a53d9f3af0f627213df77b053c06cR45 and
https://github.com/ChainSafe/gossamer/pull/2566/files#diff-3abb401d0f0c9e3a6c9f813400f674af4b2a53d9f3af0f627213df77b053c06cR53

Reply:
done